### DISCUSSION TITLE
Leadscrew pitch and backlash

### DISCUSSION THREAD

**Discussion by Festivejelly** [kachurovskiy/nanoels#142] [Q&A] [answered]
I finally got my build together (video incomming)

But I noticed while the X axis is pretty much spot on the Z axis is not and is quite a bit out.

I measure the pitch of the leadscrew on the Z axis and as far as I can tell its 2mm.

So here are my settings 

// Main lead screw (Z) parameters.
const long SCREW_Z_DU = 20000; // 2mm lead screw in deci-microns (10^-7 of a meter)
const long MOTOR_STEPS_Z = 200;
const long SPEED_START_Z = 2 * MOTOR_STEPS_Z; // Initial speed of a motor, steps / second.
const long ACCELERATION_Z = 30 * MOTOR_STEPS_Z; // Acceleration of a motor, steps / second ^ 2.
const long SPEED_MANUAL_MOVE_Z = 6 * MOTOR_STEPS_Z; // Maximum speed of a motor during manual move, steps / second.
const bool INVERT_Z = false; // change (true/false) if the carriage moves e.g. "left" when you press "right".
const bool NEEDS_REST_Z = true; // Set to false for closed-loop drivers, true for open-loop.
const long MAX_TRAVEL_MM_Z = 300; // Lathe bed doesn't allow to travel more than this in one go, 30cm / ~1 foot
const long BACKLASH_DU_Z = 6500; // 0.65mm backlash in deci-microns (10^-7 of a meter)
const char NAME_Z = 'Z'; // Text shown on screen before axis position value, GCode axis name

Could it be that the backlash is set too high or too low?

I also noticed that the RPM jumps around a little. Only by 2 - 5 RPM, I guess this is inevitable?

All in all though it seems to work super nice, I just need to get the axis moving accurately. I do have a DRO which is pretty accurate so maybe I can use that to calibrate some how?

**Comment by kachurovskiy** [comment, marked as answer]
One thing to keep in mind, especially for machines that were heavily used, is that leadscrew has variable wear in different parts so it would be a natural limitation of what can be achieved with H4 in both single and bi-directional movement accuracy. With that out of the way:

1. Ensure that moving in 1 direction gives you accurate results. E.g. set a dial indicator so that it measures movement along the desired axis, move a bit to the left, zero indicator, move `A` mm to the left using H4, check whether the dial indicator shows `A`. If it's consistently different, `SCREW_Z_DU` likely needs to be adjusted.
2. To calculate backlash, set dial indicator so that it measures movement along the desired axis, push the carriage with your hands in one direction, zero indicator, push in the other direction, the dial indicator reading is the backlash.

DRO replaces the dial indicator nicely.

**Reply by Festivejelly** [reply]
Ive barely used the leadscrew on my lathe. Its less than a year old and I always feed by hand so I dont think its worn. Might be out of alignment though. I wonder if theres room in there for a ball screw... future project maybe. I'll try re aligning everything and trying again. Was very impressed how accurate the X axis was though! Dro was showing only 0.02 deviation.

**Reply by Festivejelly** [reply]
Regarding point 2:

For measuring the backlash do I engage the half nut and use the carriage wheel to move left till it stops, zero DRO then move right till it stops then thats the backlash?

**Reply by kachurovskiy** [reply]
Yes, that's one option. Another option would be to use H4 to move 2mm to the left, zero DRO, move another 2mm to the left, then 2mm to the right. The DRO value has to be added/subtracted from the currently specified backlash so that when you repeat the process after updating the configuration, DRO shows 0.

**Reply by Festivejelly** [reply]
Okay when I move the Z and zero it seems to be moving 8.70 - 8.85mm when I ask it to move 10mm

When I move left to right its pretty much always 8.75mm when I move from right to left its between 8.7 - 8.85 probably due to play. Pretty sure my leadscrew is bent so I loosened the bearing block on the right side. Seems to move smoother now.

I did a backlash test looks like I get about 0.22mm backlash.

Is there a way to change these values on the device directly or do I need to compile the firmware again and upload it?

**Reply by kachurovskiy** [reply]
Looks like you need to change your SCREW_Z_DU to 17500, BACKLASH_DU_Z to 8700 and re-upload the firmware. That's the only way to change hardware parameters.

**Reply by Festivejelly** [reply]
So I just measured the pitch of the leadscrew and checked with the manufacturer and its definitely 2mm. So I cant get my head around why its not moving the correct amount. Im using the belts and pulleys to connect the leadscrew to the Nema 23. There doesnt seem to be a huge amount of play. Could the leadscrew be shifting left to right slightly?

**Reply by kachurovskiy** [reply]
It's hard to say what could be going on mechanically without seeing machine in operation. Perhaps you could share a short video showing continuously moving in one direction e.g. by 5mm and showing what's on the DRO after each move? Also a glimpse of the mechanics.

**Reply by Festivejelly** [reply]
Ill do a short video. Im going to do some adjustments someone suggestion on the Warco FB page first. I want to rule out any mechanical issues before I do anything because I think maybe the screw could be moving slightly left and right. So maybe I need to torque down the nuts and each end so that it puts a little more pressure on the thrust bearings. That said im a bit concerned about doing it on the stepper side as I dont want to over torque the plastic gear.

**Reply by Festivejelly** [reply]
Here is a video https://youtu.be/6VZD0DE_qlY?si=wypiY43MEmuZZ7Lj

i cant observe the leadscrew wobbling and it looks like 2mm to me…

**Reply by Festivejelly** [reply]
Also i put a dial indicator on the carriage against the chuck and its reading the same as the dro.

**Reply by kachurovskiy** [reply]
You might want to measure the leadscrew pitch in a more precise manner e.g. by measuring 10 thread turns and dividing the result by 10.

Seems like setting SCREW_Z_DU to 17500 should solve the issue.

**Reply by Festivejelly** [reply]
I measured over 35mm and there were 17 threads which equals 2. So im really stumped. I marked the leaderscrew and turned it a whole turn by using a spanner and the carriage moved 1.8mm So I wonder if my half nut is buggered.

**Comment by Festivejelly** [comment]
So something odd is going on. When I move small amounts it moves too little. But if I tell it to move 30mm it actually moves 32mm with SCREW_Z_DU = 18000

So I think something is up with the config?

Maybe I need to change this?

const long SPEED_START_Z = 2 * MOTOR_STEPS_Z; // Initial speed of a motor, steps / second.
const long ACCELERATION_Z = 30 * MOTOR_STEPS_Z; // Acceleration of a motor, steps / second ^ 2.

**Reply by Festivejelly** [reply]
I wonder if the pulleys are slipping on the shafts. Ive remixed the design to use 2 grub screws will print those and test again.

**Reply by kachurovskiy** [reply]
You can check whether the pulleys are slipping by making a mark across both pulley and the shaft.

I haven't seen a case yet where an axis would **sometimes** over-travel. If you overall see inconsistency in motor movements, consider providing additional 5V supply to your driver signal lines as per https://github.com/kachurovskiy/nanoels/tree/main/h4#wiring

You're using open-loop motors so too-high acceleration could technically result in lost steps though it's rare. If you suspect that it's happening, try lowering acceleration in half.

Otherwise, once the mechanical issues are excluded, I would adjust SCREW_Z_DU and SCREW_X_DU as per method we've discussed earlier (new `SCREW_Z_DU` = old `SCREW_Z_DU` * `REQUESTED_MM` / `DRO_MM`) and tune the backlash value.

**Reply by Festivejelly** [reply]
I think there were some mechanical issues so i stripped the whole thing down, milled extra adjustment in the holes holding the blocks that hold the leadscrew on. I also cleaned and lubed the half nut and re adjusted it. I also added a small washer on the right end of the leadscrew after the thrust bearing so that I can tighten the hex nut against it. Now there is no sideways movement and when I rotate the leadscrew 360 degrees it moves almost exactly 2mm.

So now I think mechanically it is sound.

I'll re try it these evening after wiring it all back up.

One thing I did notice now and again was the RPM was reading random values while nothing is going on. I guess this could be signal noise?

Ive since added a ferrite core to the encoder (both ends of the wire). And I plan on supplying 5v from a dedicated supply to the drivers just like I did with the encoder.

I'll do that tonight and hopefully it will yield good results.

The folks on the Warco FB page were impressed with the controller!

**Reply by Festivejelly** [reply]
So just to check my external 5v should be going to PUL+, DIR+, ENA+ on the stepper driver.

**Reply by kachurovskiy** [reply]
> One thing I did notice now and again was the RPM was reading random values while nothing is going on. I guess this could be signal noise?

This is not normal, encoder wires should be in a shielded cable with the shielding connected to GND on H4 as per https://github.com/kachurovskiy/nanoels/blob/main/h4/README.md#wiring

Also ensure that when rotating the chuck by hand exactly 1 turn forward, angle shown on screen changes by 360 degrees to the same value it was on.

> The folks on the Warco FB page were impressed with the controller!

Thanks for sharing it there.

> So just to check my external 5v should be going to PUL+, DIR+, ENA+ on the stepper driver.

Correct.

**Reply by Festivejelly** [reply]
So I crimped all of the wires to ensure a good connection and the encoder angle doesnt show. It only shows if I start the lathe and have it rotate. At first the RPM shows a few hundred over what the actual RPM is, then it settles down to the correct value after a while. If I then stop the lathe and manually rotate by hand the angle does seem to show the correct angle. So seems like it takes a little while to get going.

Im going to do some tests cutting brass to see if it moves the correct amounts. When moving the leadscrew by hand now its very accurate with very little backlash (measuring with a DRO and dial indicator). But just doing a brief test I was seeing that it was moving under the correct amount slightly, but I wonder if thats just how im testing it. So ill do some cuts and measure them.

**Reply by kachurovskiy** [reply]
> At first the RPM shows a few hundred over what the actual RPM is, then it settles down to the correct value after a while. 

This is a sign of interference from the main motor startup showing as encoder pulses. We've seen it before in a [VFD setups](https://github.com/kachurovskiy/nanoels/discussions/89). Please post pictures of your encoder wiring.

> So ill do some cuts and measure them.

You're likely to see problems with keeping the cutter in the thread if the motor startup produces ghost encoder pulses.

**Reply by Festivejelly** [reply]
After a while it settles down to the correct RPM. And then it becomes responsive. Its just for that first few rotations it seems to do nothing even with the lathe off. I'll send some pics of the wiring tomorrow. There were 4 wires and shielding.

VCC Red
Ground Black
Signal A Green
Signal B White
Shielding.

I twisted the shielding around ground and put that into the terminal.

I connected the signal wires to the ENCA and ENCB and I spliced a 5v wire from a 5v power supply the powers the Nano, and off that splice I run the 5v to the Stepper drivers and the encoder.

Here is a crude diagram of my wiring

![image](https://github.com/kachurovskiy/nanoels/assets/2408274/386ed003-45a9-4e0c-942e-1682f730d0a8)

**Reply by Festivejelly** [reply]
Also im not sure my motor is a VFD. I think its just a 550watt brushless motor.

**Reply by kachurovskiy** [reply]
@Festivejelly the diagram looks good. Here's an example of how I wire the encoder:

![PXL_20231129_075052413](https://github.com/kachurovskiy/nanoels/assets/517919/030b491b-6dca-45d3-8923-3d48c9ae03e6)

**Reply by Festivejelly** [reply]
So my wiring looks the same.

Im getting some peculiar behaviour.

If I tell it to move 10mm on the Z it moves 8.75mm. If I tell it to move 30mm it moves 28.75.... so its a constantly roughly 1,25mm error. If the pitch of the leadscrew or halfnuts was off I would expect and increasing error the longer the distance.

Also my rotary encoder doesnt seem to be working or at least its not showing movement in the controller anymore. Ive ordered a new one to be sure.

X axis still functions fine but now oddly it doesnt seem to hold anymore where as previously it did. Maybe this is because im also supplying 5v to the stepper drivers. Who knows i'll revert back and see.

Here is my config
`// Define your hardware parameters here.
const int ENCODER_STEPS_INT = 600; // 600 step spindle optical rotary encoder. Fractional values not supported.
const int ENCODER_BACKLASH = 3;    // Numer of impulses encoder can issue without movement of the spindle

// Spindle rotary encoder pins. Swap values if the rotation direction is wrong.
#define ENC_A 7
#define ENC_B 15

// Main lead screw (Z) parameters.
const long SCREW_Z_DU = 20000; // 2mm lead screw in deci-microns (10^-7 of a meter)
const long MOTOR_STEPS_Z = 200;
const long SPEED_START_Z = 2 * MOTOR_STEPS_Z;       // Initial speed of a motor, steps / second.
const long ACCELERATION_Z = 30 * MOTOR_STEPS_Z;     // Acceleration of a motor, steps / second ^ 2.
const long SPEED_MANUAL_MOVE_Z = 6 * MOTOR_STEPS_Z; // Maximum speed of a motor during manual move, steps / second.
const bool INVERT_Z = true;                         // change (true/false) if the carriage moves e.g. "left" when you press "right".
const bool NEEDS_REST_Z = true;                     // Set to false for closed-loop drivers, true for open-loop.
const long MAX_TRAVEL_MM_Z = 300;                   // Lathe bed doesn't allow to travel more than this in one go, 30cm / ~1 foot
const long BACKLASH_DU_Z = 2000;                    // 0.65mm backlash in deci-microns (10^-7 of a meter)
const char NAME_Z = 'Z';                            // Text shown on screen before axis position value, GCode axis name

// Cross-slide lead screw (X) parameters.
const long SCREW_X_DU = 10000;                      // 1.0mm lead screw with 3x reduction in deci-microns (10^-7) of a meter
const long MOTOR_STEPS_X = 600;                     // 800 steps at 3x reduction
const long SPEED_START_X = MOTOR_STEPS_X;           // Initial speed of a motor, steps / second.
const long ACCELERATION_X = 10 * MOTOR_STEPS_X;     // Acceleration of a motor, steps / second ^ 2.
const long SPEED_MANUAL_MOVE_X = 3 * MOTOR_STEPS_X; // Maximum speed of a motor during manual move, steps / second.
const bool INVERT_X = false;                        // change (true/false) if the carriage moves e.g. "left" when you press "right".
const bool NEEDS_REST_X = false;                    // Set to false for all kinds of drivers or X will be unlocked when not moving.
const long MAX_TRAVEL_MM_X = 100;                   // Cross slide doesn't allow to travel more than this in one go, 10cm
const long BACKLASH_DU_X = 1500;                    // 0.15mm backlash in deci-microns (10^-7 of a meter)
const char NAME_X = 'X';                            // Text shown on screen before axis position value, GCode axis name

// Manual stepping with left/right/up/down buttons. Only used when step isn't default continuous (1mm or 0.1").
const long STEP_TIME_MS = 500;          // Time in milliseconds it should take to make 1 manual step.
const long DELAY_BETWEEN_STEPS_MS = 80; // Time in milliseconds to wait between steps.`

Another thing ive noticed and ive taken a video of this which ill upload. If I tell the Z stepper to move left the stepper rotates the other way just a tiny bit before going left. I assumed this was backlash compensation or something but that doesnt really make sense to do if you are not changing direction. given the backlash is set as 2000 as thats what ive measured it to be maybe this is the source of that 1.25mm difference I was talking about. Seems odd to me the motor would move the opposite way before moving the correct way if im not changing direction.

**Comment by Festivejelly** [comment]
Video showing the stepper issue

https://youtu.be/QGqJmdaca8U?si=8pGdsvehPK5PJ97-

this might explain why its always the same amount out regardless of the length it moves.

**Reply by kachurovskiy** [reply]
Thanks for filming and sharing. I think the movement is due to `NEEDS_REST_Z` being `true` which means that stepper disables and re-enables between every move. It seems like enabling causes it to move. Try setting `NEEDS_REST_Z` to `false`.

**Reply by Festivejelly** [reply]
@kachurovskiy Ive read some reviews on the encoder that im using might need some pull up resistors as the signal is below 5v. Is this something you've experienced? Ive ordered a new encoder of a similar type, but I bet they all come from the same place.

**Reply by kachurovskiy** [reply]
Most encoders need pull-up resistors on signal lines, I've got them integrated into the H4 PCB already.

**Reply by Festivejelly** [reply]
@kachurovskiy setting needs_rest_z to false fixed it! Thanks for all the support. It works like a charm now and ive dialled it in to only need 0.02mm backlash! The halfnuts must be well machined!

**Reply by Festivejelly** [reply]
Measured the voltage from my 5v supply to the encoder… its dropping below 5vs. Ive ordered a dedicated supply and will use a booster if needed. Afaik its 5-24 so ill maybe set only the encoder to take 6v
